### ISSUE TITLE
SDK tests "fail" too often due to open handles

### ISSUE THREAD

**Issue by kelsos** [raiden-network/light-client#925]
## Description

A lot of times lately the pull requests fail at the SDK tests. On close observation, we can usually see that the tests pass successfully but jest fails to terminate due to some async tasks not completing properly. 

Since this started happening way too often lately I decided to investigate using the suggested `--detectOpenHandles` jest flag. After sshing into the CI and running the tests in what seemed forever I managed to get the following output:

```javascript
  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "sent closeChannel tx "0xd0c03071074f4a7a7c9c7a133a473d98acdee4f0b4b15526fedffcb2b0181aa9" to "0x3979Aa776fE2ef327dFD0747aF6F4EF5A0EE4517"".



      at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:183:10)
      at TapSubscriber._tapNext (src/channels/epics.ts:7744:22)
      at TapSubscriber.Object.<anonymous>.TapSubscriber._next (node_modules/rxjs/src/internal/operators/tap.ts:120:21)
      at TapSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at MergeMapSubscriber.Object.<anonymous>.MergeMapSubscriber.notifyNext (node_modules/rxjs/src/internal/operators/mergeMap.ts:168:22)
      at InnerSubscriber.Object.<anonymous>.InnerSubscriber._next (node_modules/rxjs/src/internal/InnerSubscriber.ts:17:17)
      at InnerSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at node_modules/rxjs/src/internal/util/subscribeToPromise.ts:8:20
          at runMicrotasks (<anonymous>)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "closeChannel tx "0xd0c03071074f4a7a7c9c7a133a473d98acdee4f0b4b15526fedffcb2b0181aa9" successfuly mined!".



      at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:183:10)
      at MapSubscriber.project (src/channels/epics.ts:7778:15)
      at MapSubscriber.Object.<anonymous>.MapSubscriber._next (node_modules/rxjs/src/internal/operators/map.ts:84:29)
      at MapSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at MergeMapSubscriber.Object.<anonymous>.MergeMapSubscriber.notifyNext (node_modules/rxjs/src/internal/operators/mergeMap.ts:168:22)
      at InnerSubscriber.Object.<anonymous>.InnerSubscriber._next (node_modules/rxjs/src/internal/InnerSubscriber.ts:17:17)
      at InnerSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at node_modules/rxjs/src/internal/util/subscribeToPromise.ts:8:20
          at runMicrotasks (<anonymous>)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "mining 21 blocks after blockNumber=49".

      45 |   public async mine(count = 1): Promise<number> {
      46 |     const blockNumber = await this.getBlockNumber();
    > 47 |     console.debug(`mining ${count} blocks after blockNumber=${blockNumber}`);
         |             ^
      48 |     const promise = new Promise(resolve => {
      49 |       const cb = (b: number): void => {
      50 |         if (b < blockNumber + count) return;

      at CustomConsole.debug (node_modules/@jest/console/build/CustomConsole.js:135:10)
      at TestProvider.<anonymous> (tests/e2e/provider.ts:47:13)
      at fulfilled (tests/e2e/provider.ts:5:58)
          at runMicrotasks (<anonymous>)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "sent settleChannel tx "0x1692fe03aabf348595d9dc1a5534bfb5f5533b40881a21dddc82afdd3cf08c70" to "0x3979Aa776fE2ef327dFD0747aF6F4EF5A0EE4517"".



      at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:183:10)
      at TapSubscriber._tapNext (src/channels/epics.ts:7940:22)
      at TapSubscriber.Object.<anonymous>.TapSubscriber._next (node_modules/rxjs/src/internal/operators/tap.ts:120:21)
      at TapSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at node_modules/rxjs/src/internal/util/subscribeToPromise.ts:8:20
          at runMicrotasks (<anonymous>)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "settleChannel tx "0x1692fe03aabf348595d9dc1a5534bfb5f5533b40881a21dddc82afdd3cf08c70" successfuly mined!".



      at CustomConsole.log (node_modules/@jest/console/build/CustomConsole.js:183:10)
      at MapSubscriber.project (src/channels/epics.ts:7974:15)
      at MapSubscriber.Object.<anonymous>.MapSubscriber._next (node_modules/rxjs/src/internal/operators/map.ts:84:29)
      at MapSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at MergeMapSubscriber.Object.<anonymous>.MergeMapSubscriber.notifyNext (node_modules/rxjs/src/internal/operators/mergeMap.ts:168:22)
      at InnerSubscriber.Object.<anonymous>.InnerSubscriber._next (node_modules/rxjs/src/internal/InnerSubscriber.ts:17:17)
      at InnerSubscriber.Object.<anonymous>.Subscriber.next (node_modules/rxjs/src/internal/Subscriber.ts:99:12)
      at node_modules/rxjs/src/internal/util/subscribeToPromise.ts:8:20
          at runMicrotasks (<anonymous>)
```

At least from what I can guess is that in some cases some of the e2e on-chain actions don't seem to properly complete before the end of the test suite.
## Acceptance criteria
- 

## Tasks
- [ ]

**Comment by kelsos** [comment]
![image](https://user-images.githubusercontent.com/2269732/73064319-449cb980-3ea1-11ea-9d4d-ea9566e02ae7.png) This has been the case this morning. 

Every single time all the tests pass but jest fails to terminate.
```bash
Test Suites: 12 passed, 12 total
Tests:       286 passed, 286 total
Snapshots:   0 total
Time:        286.739s
Ran all test suites.
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
Too long with no output (exceeded 10m0s): context deadline exceeded
```

I am running the tests in docker trying to replicate by using `circleci:node:12-browsers` image but I didn't have any success so far.

The command looks like this:
```bash
docker run --cpus=0.6 -m 512MB -it -v /home/kelsos/development/github/raiden-network/light-client/raiden-ts:/sdk circleci/node:12-browsers /bin/sh -c cd /sdk; npm run test -- --ci --runInBand
```

The next step is to clone the repo in a new folder and have it run on repeat in hope of reproducing the same result.

**Comment by kelsos** [comment]
So running the tests in a loop helps to get the issue at some point.

**Comment by kelsos** [comment]
On Friday I manage to get a failure in less than 10 runs running all the tests in this constrained docker image. Circle CI uses 2 CPUs and 4GB for RAM for the docker image that runs our test suite. 

Initially, when I was falling to replicate the flakiness I was running with a `--silent` flag in jest that would suppress all logging output. 

I am now running again 50 rounds of `--silent` to see if what happened was pure chance or if the verbose logging output might somehow affect jest. 

It could be that jest has trouble processing all this output and the logging continues after the test completion causing jest termination to fail. At least that is the hypothesis I am currently trying to verify.

After completing 50 runs with `--silent` I will also run another 50 without to see how fast I will run into the issue.

**Comment by kelsos** [comment]
It seems that it managed to fail after 2h30m of running even with the output set to silent so this seems to disprove my hypothesis. I will continue to run some tests on the background to see if I get some more information

**Comment by andrevmatos** [comment]
I've fixed this in d831b0feae84f593d17550e48b9565eb78c2ae89 as part of #975 .
The timeout problem seems to be either some small performance regression with some past dependency update (maybe jest, or ganache-cli), or further environment constraint of CircleCI.
This was more evident by the fact that tests passed locally for me and kelsos and he could only reproduce it by hammering/constraing the test environment, triggering the timeouts.
In any case, just giving a little more time for the tests seems to have fixed it, as rebased PRs #973 #974 & #976 tests passed on first run.
Let's reopen if issue reappears.

**Comment by kelsos** [comment]
@andrevmatos great, thank you for taking care of it :)
